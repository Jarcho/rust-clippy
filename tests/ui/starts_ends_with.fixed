#![allow(clippy::needless_if, dead_code, unused_must_use)]

fn main() {}

#[allow(clippy::unnecessary_operation)]
fn starts_with() {
    "".starts_with(' '); //~ chars_next_cmp
    !"".starts_with(' '); //~ chars_next_cmp

    // Ensure that suggestion is escaped correctly
    "".starts_with('\n'); //~ chars_next_cmp
    !"".starts_with('\n'); //~ chars_next_cmp
}

fn chars_cmp_with_unwrap() {
    let s = String::from("foo");
    //~v chars_next_cmp
    if s.starts_with('f') {
        // s.starts_with('f')
        // Nothing here
    }
    //~v chars_last_cmp
    if s.ends_with('o') {
        // s.ends_with('o')
        // Nothing here
    }
    //~v chars_last_cmp
    if s.ends_with('o') {
        // s.ends_with('o')
        // Nothing here
    }
    //~v chars_next_cmp
    if !s.starts_with('f') {
        // !s.starts_with('f')
        // Nothing here
    }
    //~v chars_last_cmp
    if !s.ends_with('o') {
        // !s.ends_with('o')
        // Nothing here
    }
    //~v chars_last_cmp
    if !s.ends_with('\n') {
        // !s.ends_with('o')
        // Nothing here
    }
}

#[allow(clippy::unnecessary_operation)]
fn ends_with() {
    "".ends_with(' '); //~ chars_last_cmp
    !"".ends_with(' '); //~ chars_last_cmp
    "".ends_with(' '); //~ chars_last_cmp
    !"".ends_with(' '); //~ chars_last_cmp

    // Ensure that suggestion is escaped correctly
    "".ends_with('\n'); //~ chars_last_cmp
    !"".ends_with('\n'); //~ chars_last_cmp
}
