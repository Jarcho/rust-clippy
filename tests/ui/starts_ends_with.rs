#![allow(clippy::needless_if, dead_code, unused_must_use)]

fn main() {}

#[allow(clippy::unnecessary_operation)]
fn starts_with() {
    "".chars().next() == Some(' '); //~ chars_next_cmp
    Some(' ') != "".chars().next(); //~ chars_next_cmp

    // Ensure that suggestion is escaped correctly
    "".chars().next() == Some('\n'); //~ chars_next_cmp
    Some('\n') != "".chars().next(); //~ chars_next_cmp
}

fn chars_cmp_with_unwrap() {
    let s = String::from("foo");
    //~v chars_next_cmp
    if s.chars().next().unwrap() == 'f' {
        // s.starts_with('f')
        // Nothing here
    }
    //~v chars_last_cmp
    if s.chars().next_back().unwrap() == 'o' {
        // s.ends_with('o')
        // Nothing here
    }
    //~v chars_last_cmp
    if s.chars().last().unwrap() == 'o' {
        // s.ends_with('o')
        // Nothing here
    }
    //~v chars_next_cmp
    if s.chars().next().unwrap() != 'f' {
        // !s.starts_with('f')
        // Nothing here
    }
    //~v chars_last_cmp
    if s.chars().next_back().unwrap() != 'o' {
        // !s.ends_with('o')
        // Nothing here
    }
    //~v chars_last_cmp
    if s.chars().last().unwrap() != '\n' {
        // !s.ends_with('o')
        // Nothing here
    }
}

#[allow(clippy::unnecessary_operation)]
fn ends_with() {
    "".chars().last() == Some(' '); //~ chars_last_cmp
    Some(' ') != "".chars().last(); //~ chars_last_cmp
    "".chars().next_back() == Some(' '); //~ chars_last_cmp
    Some(' ') != "".chars().next_back(); //~ chars_last_cmp

    // Ensure that suggestion is escaped correctly
    "".chars().last() == Some('\n'); //~ chars_last_cmp
    Some('\n') != "".chars().last(); //~ chars_last_cmp
}
