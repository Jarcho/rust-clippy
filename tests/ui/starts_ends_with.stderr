error: you should use the `starts_with` method
  --> $DIR/starts_ends_with.rs:7:5
   |
LL |     "".chars().next() == Some(' ');
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: like this: `"".starts_with(' ')`
   |
   = note: `-D clippy::chars-next-cmp` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::chars_next_cmp)]`

error: you should use the `starts_with` method
  --> $DIR/starts_ends_with.rs:8:5
   |
LL |     Some(' ') != "".chars().next();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: like this: `!"".starts_with(' ')`

error: you should use the `starts_with` method
  --> $DIR/starts_ends_with.rs:11:5
   |
LL |     "".chars().next() == Some('\n');
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: like this: `"".starts_with('\n')`

error: you should use the `starts_with` method
  --> $DIR/starts_ends_with.rs:12:5
   |
LL |     Some('\n') != "".chars().next();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: like this: `!"".starts_with('\n')`

error: you should use the `starts_with` method
  --> $DIR/starts_ends_with.rs:18:8
   |
LL |     if s.chars().next().unwrap() == 'f' {
   |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: like this: `s.starts_with('f')`

error: you should use the `ends_with` method
  --> $DIR/starts_ends_with.rs:23:8
   |
LL |     if s.chars().next_back().unwrap() == 'o' {
   |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: like this: `s.ends_with('o')`
   |
   = note: `-D clippy::chars-last-cmp` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::chars_last_cmp)]`

error: you should use the `ends_with` method
  --> $DIR/starts_ends_with.rs:28:8
   |
LL |     if s.chars().last().unwrap() == 'o' {
   |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: like this: `s.ends_with('o')`

error: you should use the `starts_with` method
  --> $DIR/starts_ends_with.rs:33:8
   |
LL |     if s.chars().next().unwrap() != 'f' {
   |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: like this: `!s.starts_with('f')`

error: you should use the `ends_with` method
  --> $DIR/starts_ends_with.rs:38:8
   |
LL |     if s.chars().next_back().unwrap() != 'o' {
   |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: like this: `!s.ends_with('o')`

error: you should use the `ends_with` method
  --> $DIR/starts_ends_with.rs:43:8
   |
LL |     if s.chars().last().unwrap() != '\n' {
   |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: like this: `!s.ends_with('\n')`

error: you should use the `ends_with` method
  --> $DIR/starts_ends_with.rs:51:5
   |
LL |     "".chars().last() == Some(' ');
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: like this: `"".ends_with(' ')`

error: you should use the `ends_with` method
  --> $DIR/starts_ends_with.rs:52:5
   |
LL |     Some(' ') != "".chars().last();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: like this: `!"".ends_with(' ')`

error: you should use the `ends_with` method
  --> $DIR/starts_ends_with.rs:53:5
   |
LL |     "".chars().next_back() == Some(' ');
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: like this: `"".ends_with(' ')`

error: you should use the `ends_with` method
  --> $DIR/starts_ends_with.rs:54:5
   |
LL |     Some(' ') != "".chars().next_back();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: like this: `!"".ends_with(' ')`

error: you should use the `ends_with` method
  --> $DIR/starts_ends_with.rs:57:5
   |
LL |     "".chars().last() == Some('\n');
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: like this: `"".ends_with('\n')`

error: you should use the `ends_with` method
  --> $DIR/starts_ends_with.rs:58:5
   |
LL |     Some('\n') != "".chars().last();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: like this: `!"".ends_with('\n')`

error: aborting due to 16 previous errors

